### ISSUE TITLE
Tree based explanations

### ISSUE THREAD

**Issue by NeverNude** [marcotcr/lime#101]
Hey there,

I was thinking about adding Tree based explanations to LIME.  It seems like a lot of the explanations assuming that we're getting linear results.  I was curious how you would tackle refactoring this to make the format of explanations more general as opposed to assuming it was linear.  I was going to add subclasses to explanations.py for TreeExplanations and another for LinearExplanations, which I think is likely uncontroversial.

However, I'm unsure of the preferred way of changing the format for explain_instance in lime_tabular for example.  There are a couple of options
1) Return dummy intercept and local_exp entries, and augment with additional entries for tree explanation.  The interface stays consistent for old implementation, and new clients have to differentiate the results.
2) Rather than returning intercept and local_exp as top level entries in the response, nest them under explanation_params which are conditional given the type of explanation that's generated (I don't know if there are plans or desires to support explanations of more than linear + tree types, but I suppose it's possible).  This future proofs it to make it easier to add new types of explanations.  This has the drawback of being a break change, but more consistent for clients implementing the results.

I also thought it might be nice to add some examples longer term, visualizing it, using something like http://bl.ocks.org/pprett/raw/3813537/.

Thoughts?

**Comment by marcotcr** [comment]
Yeah, subclasses to explanations.py sounds like the right approach. We should have done this already when regression was added.

Note that lime_tabular returns an explanation.Explanation - as long as linear explanations remain as the default, the interface can remain the same as long as LinearExplanations implements the current functions in the Explanation class. Thus, there is no need to have dummy intercept and local exp.

Before refactoring LIME code, I would suggest trying out some tree explanations. I have tried them multiple times and in multiple ways and they always seem too confusing, and hard for the user to parse. I have yet to see a case where one would prefer them (I'm happy to be shown otherwise).
